File: 3DRadSpace/3DRadSpace/3DRadSpaceDll/Camera.cs
using Microsoft.Xna.Framework;
using Microsoft.Xna.Framework.Graphics;
using System;
using System.Collections.Generic;
using System.Linq;
using System.Text;
using System.Threading.Tasks;

namespace _3DRadSpaceDll
{
    public class Camera : GameObject
    {
        public Camera()
        {
            Name = "Camera";
            Enabled = false;
            _pos = Vector3.Zero;
            _rot = Vector3.UnitZ;
            _fov = MathHelper.ToRadians(45);
            _far = 500f;
            _near = 0.01f;
            _aspect_ratio = 600 / 480f;
            _up = Vector3.Up;
            RecalculateMatrices();
        }
        /// <summary>
        /// View matrix
        /// </summary>
        public Matrix View;
        /// <summary>
        /// Projection
        /// </summary>
        public Matrix Projection;
        /// <summary>
        /// Position of the camera object.
        /// </summary>
        public new Vector3 Position
        {
            get
            {
                return _pos;
            }
            set
            {
                View = Matrix.CreateLookAt(value, value + _rot, _up);
                _pos = value;
            }
        }
        /// <summary>
        /// Describes a normalized vector describing the 3D rotation in unit-angles
        /// </summary>
        public new Vector3 Rotation
        {
            get
            {
                return _rot;
            }
            set
            {
                View = Matrix.CreateLookAt(_pos, value, _up);
            }
        }
        Vector3 _pos;
        Vector3 _rot;
        /// <summary>
        /// The Model used in the Editor
        /// </summary>
        public static Model EditorModel;
        /// <summary>
        /// Describes the camera's up direction as a unit vector.
        /// </summary>
        public Vector3 Up
        {
            get
            {
                return _up;
            }
            set
            {
                View = Matrix.CreateLookAt(_pos, _rot, value);
                _up = value;
            }
        }
        Vector3 _up;
        /// <summary>
        /// Field of view in radians.
        /// </summary>
        public float FOV
        {
            get
            {
                return _fov;
            }
            set
            {
                Projection = Matrix.CreatePerspectiveFieldOfView(value, _aspect_ratio, _near, _far);
                _fov = value;
            }
        }
        float _fov;
        /// <summary>
        /// Aspect ratio defined as x/y px .
        /// </summary>
        public float AspectRatio
        {
            get
            {
                return _aspect_ratio;
            }
            set
            {
                Projection = Matrix.CreatePerspectiveFieldOfView(_fov, value, _near, _far);
                _aspect_ratio = value;
            }
        }
        float _aspect_ratio;
        /// <summary>
        /// Minimum drawing distance
        /// </summary>
        public float NearDrawDistance
        {
            get
            {
                return _near;
            }
            set
            {
                Projection = Matrix.CreatePerspectiveFieldOfView(_fov, _aspect_ratio, value,_far);
                _near = value;
            }
        }
        float _near;
        /// <summary>
        /// Maximum draw distance
        /// </summary>
        public float FarDrawDistance
        {
            get
            {
                return _far;
            }
            set
            {
                Projection = Matrix.CreatePerspectiveFieldOfView(_fov, _aspect_ratio, _near, value);
                _far = value;
            }
        }
        float _far;
        public Camera(string name,bool enabled,Vector3 pos,Vector3 rot,Vector3 up,float fov,float aspect_ratio,float near,float far)
        {
            Vector3 n = Vector3.Transform(Vector3.UnitZ, Matrix.CreateFromYawPitchRoll(rot.Y, rot.X, rot.Y));
            View = Matrix.CreateLookAt(pos, pos + n, up);
            Projection = Matrix.CreatePerspectiveFieldOfView(fov, aspect_ratio, near, far);
            Name = name;
            Enabled = enabled;
            _pos = pos;
            _rot = n;
            _up = up;
            _fov = fov;
            _aspect_ratio = aspect_ratio;
            _far = far;
            _near = near;
        }
        public BoundingFrustum Frustum
        {
            get
            {
                return new BoundingFrustum(View * Projection);
            }
        }
        public void RecalculateMatrices()
        {
            View = Matrix.CreateLookAt(_pos, _pos + _rot,_up);
            Projection = Matrix.CreatePerspectiveFieldOfView(_fov, _aspect_ratio, _near, _far);
        }
    }
}

File: 3DRadSpace/Engine3DRadSpace/Objects/Camera.cpp
#include "Camera.hpp"
#include "../Reflection/Reflection.hpp"
#include "../Graphics/Model3D.hpp"

using namespace Engine3DRadSpace::Graphics;
using namespace Engine3DRadSpace::Objects;
using namespace Engine3DRadSpace::Math;

Camera::Camera(const std::string& name,const std::string &tag, bool visible, Vector3 pos, Vector3 look_at, Vector3 up,float aspectRatio, float fov, float npd, float fpd):
	IObject3D(name,tag, visible, true, pos, Vector3::Zero(), Quaternion(), Vector3::One()),
	UpwardsDir(up),
	AspectRatio(aspectRatio),
	FieldOfView(fov),
	FarPlaneDistance(fpd),
	NearPlaneDistance(npd),
	LookAt(0,0,0),
	LookMode(CameraMode::UseRotation)
{
}

void Camera::EditorInitialize()
{
}

static std::unique_ptr<Model3D> cameraModel;

void Camera::EditorLoad(Content::ContentManager *content)
{
	if(cameraModel == nullptr)
	{
		cameraModel = std::make_unique<Model3D>(content->GetDevice(), "Data\\Models\\Camera.x");
	}
}

void Camera::Draw(Matrix4x4& view, Matrix4x4& projection, double dt)
{
	Vector3 focus(0, 0, 0);
	switch (this->LookMode)
	{
	case CameraMode::UseRotation:
		focus = Position + Vector3::UnitZ().Transform(Rotation);
		break;
	case CameraMode::UseLookAtCoordinates:
		focus = this->LookAt;
		break;
	default: //Normally we wouldn't get here. Do not set the focus vector.
		break;
	}
	
	view = Matrix4x4::CreateLookAtView(Position, focus, UpwardsDir);
	projection = Matrix4x4::CreatePerspectiveProjection(AspectRatio, FieldOfView, NearPlaneDistance, FarPlaneDistance);
}

void Camera::Update(Input::Keyboard& keyboard, Input::Mouse& mouse, double dt) { } //Do nothing

Matrix4x4 Camera::GetModelMartix()
{
	return Matrix4x4::CreateFromQuaternion(Rotation) * Matrix4x4::CreateTranslation(Position);
}

void Camera::EditorDraw(const Matrix4x4& view, const Matrix4x4& projection, double dt, bool selected)
{
	cameraModel->Draw(GetModelMartix(), view, projection);
}

std::optional<float> Camera::Intersects(const Ray&r)
{
	return r.Intersects(BoundingSphere(Position,1.5f));
}

Engine3DRadSpace::Reflection::UUID Camera::GetUUID()
{
	// {84376082-A56E-49E6-B095-BAEEF4BF29B5}
	return {0x84376082, 0xa56e, 0x49e6, { 0xb0, 0x95, 0xba, 0xee, 0xf4, 0xbf, 0x29, 0xb5 }};
}

void Camera::Initialize()
{
}

void Camera::Load(Content::ContentManager *content)
{
}

void Camera::Load(Content::ContentManager* content, const std::filesystem::path& path)
{
}

Camera::~Camera()
{
	//Deallocate camera model early, to avoid DirectX debug layer STATE_CREATION warnings, despite the Camera editor model being a std::unique_ptr.
	if (cameraModel != nullptr)
		cameraModel.reset();
}

REFL_BEGIN(Camera,"Camera","Camera objects","Perspective camera")
	REFL_FIELD(Camera, std::string, Name, "Name", "Camera", "Camera's name")
	REFL_FIELD(Camera, bool, Enabled, "Enabled", true, "Does the camera render the scene?")
	REFL_FIELD(Camera, Vector3, Position, "Position", Vector3::Zero(), "Camera position")
	REFL_FIELD(Camera, Quaternion, Rotation, "Rotation", Quaternion(), "Camera rotation")
	REFL_FIELD(Camera, Vector3, UpwardsDir, "Upwards direction", Vector3::One(), "Camera surface normal vector")
	REFL_FIELD(Camera, float, AspectRatio, "Aspect ratio", 4.f / 3.f, "Camera aspect ratio")
	REFL_FIELD(Camera, float, FieldOfView, "Field of view", 65.f, "Camera's field of view")
	REFL_FIELD(Camera, float, NearPlaneDistance, "Near plane distance", 0.01f, "Minimum drawing distance")
	REFL_FIELD(Camera, float, FarPlaneDistance, "Far plane distance", 500.f, "Maximum drawing distance")
REFL_END